File: 1563-stone-game-v/1563-stone-game-v.cpp
class Solution {
public:
    int stoneGameV(vector<int>& stoneValue) {
        int n = stoneValue.size();
        vector<int> sum(n + 1);
        partial_sum(stoneValue.begin(), stoneValue.end(), sum.begin() + 1);
        vector<int> f(n + 1);
        vector suf_max(n + 1, vector<int>(n + 1, INT_MIN));

        for (int i = n - 1; i >= 0; i--) {
            suf_max[i][i + 1] = -sum[i];
            int pre_max = 0;
            int k = i + 1;
            for (int j = i + 2; j <= n; j++) {
                while (sum[k] - sum[i] <= sum[j] - sum[k]) {
                    pre_max = max(pre_max, f[k] + sum[k]);
                    k++;
                }
                int q = sum[k - 1] - sum[i] == sum[j] - sum[k - 1] ? k - 1 : k;
                f[j] = max(pre_max - sum[i], suf_max[q][j] + sum[j]);
                suf_max[i][j] = max(suf_max[i + 1][j], f[j] - sum[i]);
            }
        }

        return f[n];
    }
};

File: 1563-stone-game-v/1563-stone-game-v.java
class Solution {
    public int stoneGameV(int[] stoneValue) {
        int n = stoneValue.length;

        long[] prefix = new long[n + 1];

        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + stoneValue[i];
        }

        int[][] dp = new int[n][n];

        int[][] leftBest = new int[n][n];

        int[][] rightBest = new int[n][n];

        int[] leftPtr = new int[n];

        int[] rightPtr = new int[n];

        for (int i = 0; i < n; i++) {
            leftBest[i][i] = stoneValue[i];
            rightBest[i][i] = stoneValue[i];

            leftPtr[i] = i - 1;

            rightPtr[i] = i;
        }

        for (int len = 2; len <= n; len++) {
            for (int l = 0; l + len <= n; l++) {
                int r = l + len - 1;

                long total = prefix[r + 1] - prefix[l];

                while (leftPtr[l] + 1 <= r - 1) {
                    int k = leftPtr[l] + 1;
                    long leftSum = prefix[k + 1] - prefix[l];

                    if (2 * leftSum > total) {
                        break;
                    }

                    leftPtr[l]++;
                }

                while (rightPtr[l] <= r - 1) {
                    int k = rightPtr[l];
                    long leftSum = prefix[k + 1] - prefix[l];

                    if (2 * leftSum >= total) {
                        break;
                    }

                    rightPtr[l]++;
                }

                int best = 0;

                if (leftPtr[l] >= l) {
                    best = leftBest[l][leftPtr[l]];
                }

                if (rightPtr[l] <= r - 1) {
                    best = Math.max(best, rightBest[rightPtr[l] + 1][r]);
                }

                dp[l][r] = best;

                leftBest[l][r] = Math.max(
                    leftBest[l][r - 1],
                    dp[l][r] + (int) total
                );

                rightBest[l][r] = Math.max(
                    rightBest[l + 1][r],
                    dp[l][r] + (int) total
                );
            }
        }

        return dp[0][n - 1];
    }
}